tests/test_grpc_server.py

```diff
@@ -71,6 +73,263 @@ def test_read_object(self):
+        content = media[QUANTUM : 2 * QUANTUM]
+        r2 = storage_pb2.WriteObjectRequest(
+            write_offset=offset,
+            checksummed_data=storage_pb2.ChecksummedData(
+                content=content, crc32c=crc32c.crc32c(content)
+            ),
+            finish_write=False,
+        )
+
+        offset = 2 * QUANTUM
+        content = media[QUANTUM:]
+        r3 = storage_pb2.WriteObjectRequest(
+            write_offset=QUANTUM,
+            checksummed_data=storage_pb2.ChecksummedData(
+                content=content, crc32c=crc32c.crc32c(content)
+            ),
+            finish_write=True,
+        )
+
+        write = self.grpc.WriteObject([r1, r2, r3], "fake-context")
+        self.assertIsNotNone(write)
+        self.assertIsNotNone(write.resource)
+        blob = write.resource
+        self.assertEqual(blob.name, "object-name", msg=write)
+        self.assertEqual(blob.bucket, "projects/_/buckets/bucket-name")
+
+    def test_object_write_invalid_request(self):
+        context = unittest.mock.Mock()
+        context.abort = unittest.mock.MagicMock()
+        write = self.grpc.WriteObject([], context)
+        context.abort.assert_called_once()
+        self.assertIsNone(write)
+
+    def test_object_write_incomplete(self):
+        QUANTUM = 256 * 1024
+        media = TestGrpc._create_block(2 * QUANTUM + QUANTUM / 2).encode("utf-8")
+
+        offset = 0
+        content = media[0:QUANTUM]
+        r1 = storage_pb2.WriteObjectRequest(
+            write_object_spec=storage_pb2.WriteObjectSpec(
+                resource={
+                    "name": "object-name",
+                    "bucket": "projects/_/buckets/bucket-name",
+                },
+            ),
+            write_offset=offset,
+            checksummed_data=storage_pb2.ChecksummedData(
+                content=content, crc32c=crc32c.crc32c(content)
+            ),
+            finish_write=False,
+        )
+
+        context = unittest.mock.Mock()
+        context.abort = unittest.mock.MagicMock()
+        write = self.grpc.WriteObject([r1], context)
+        context.abort.assert_called_once()
+        self.assertIsNone(write)
+
+    def test_resumable_write(self):
+        start = self.grpc.StartResumableWrite(
+            storage_pb2.StartResumableWriteRequest(
+                write_object_spec=storage_pb2.WriteObjectSpec(
+                    resource=storage_pb2.Object(
+                        name="object-name", bucket="projects/_/buckets/bucket-name"
+                    )
+                )
+            ),
+            context="fake-context",
+        )
+        self.assertIsNotNone(start.upload_id)
+
+        def streamer():
+            media = TestGrpc._create_block(517 * 1024).encode("utf-8")
+            id = start.upload_id
+            step = storage_pb2.ServiceConstants.Values.MAX_READ_CHUNK_BYTES
+            for offset in range(0, len(media), step):
+                upload_id = id
+                id = None
```

Comment:
What is the use for `id`?

Reply:
Only the first `WriteObjectRequest` can have a `upload_id` field set.  So, after the first iteration `id` is `None`, and therefore only on the first iteration `upload_id` has a valid value.  Terrible?  Maybe.

Is there a more elegant way to yield one value that is "special" and then yield other values?  Oh... can I yield before entering the loop?

Reply:
Not sure I follow... why is it that we only want the first `WriteObjectRequest` to have a valid `upload_id`?

Reply:
I think the proto spec requires that:

https://github.com/googleapis/googleapis/blob/68e5bf3045d7b3ef55f473494a0cd4566491f899/google/storage/v2/storage.proto#L222-L232

we may not be enforcing it though.

Reply:
I change the code to yield once before entering the loop and added a comment.

Reply:
I see! Thanks for the change. Another question on line 195 :)